### PR TITLE
proxy: add local-proxy to compute image

### PR DESCRIPTION
**Pull request by conradludgate**
1. Adds local-proxy to compute image and vm spec
2. Updates local-proxy config processing, writing PID to a file eagerly
3. Updates compute-ctl to understand local proxy compute spec and to send SIGHUP to local-proxy over that pid.

closes https://github.com/neondatabase/cloud/issues/16867

### CODE REVIEW

File: vm-image-spec.yaml [hunk 1]
```diff
@@ -15,6 +15,10 @@ commands:
     user: postgres
     sysvInitAction: respawn
     shell: '/usr/local/bin/pgbouncer /etc/pgbouncer.ini'
+  - name: local-proxy
+    user: postgres
```
Comment (conradludgate) [inline vm-image-spec.yaml:19]: TODO: should this be the postgres user? local-proxy works a bit like pgbouncer in a sense, which also uses the postgres user.

File: compute_tools/src/compute.rs [hunk 2]
```diff
@@ -970,6 +994,13 @@ impl ComputeNode {
 
         self.pg_reload_conf()?;
 
+        if let Some(handle) = local_proxy_handle {
+            match handle.join() {
+                Ok(res) => res?,
+                Err(panic) => std::panic::resume_unwind(panic),
+            }
+        }
+
```
Comment (conradludgate) [inline compute_tools/src/compute.rs:1003]: Should we wait here? pgbouncer config doesn't wait. unsure.

File: compute_tools/src/compute.rs [hunk 3]
```diff
@@ -928,6 +937,21 @@ impl ComputeNode {
             });
         }
 
+        let mut local_proxy_handle = None;
```
Comment (cloneable) [inline compute_tools/src/compute.rs:940]: nit: prefer `let local_proxy_handle = if ...`

File: proxy/src/bin/local_proxy.rs [hunk 4]
```diff
@@ -139,12 +149,28 @@ async fn main() -> anyhow::Result<()> {
         16,
     ));
 
-    refresh_config(args.config_path.clone()).await;
+    // write the process ID to a file so that compute-ctl can find our process later
+    // in order to trigger the appropriate SIGHUP on config change.
+    let pid = nix::unistd::getpid();
```
Comment (cloneable) [inline proxy/src/bin/local_proxy.rs:154]: We can't use `std::process::id()`?

File: compute_tools/src/config.rs [hunk 5]
```diff
@@ -139,3 +142,38 @@ where
 
     res
 }
+
+/// Create or completely rewrite configuration file specified by `path`
+pub fn write_local_proxy_conf(path: &Path, local_proxy: &LocalProxySpec) -> Result<()> {
+    let config =
+        serde_json::to_string_pretty(local_proxy).context("serializing local-proxy json")?;
```
Comment (cloneable) [inline compute_tools/src/config.rs:149]: nit: Technically, we don't know the name of the file. Maybe replace file name with `LocalProxySpec`.

**Review by cloneable — APPROVED** [review]

File: proxy/src/bin/local_proxy.rs (hunk 4 above)
Comment (conradludgate) [reply proxy/src/bin/local_proxy.rs:154]: TIL

File: proxy/src/bin/local_proxy.rs [hunk 6]
```diff
@@ -161,12 +161,6 @@ async fn main() -> anyhow::Result<()> {
         16,
     ));
 
-    // write the process ID to a file so that compute-ctl can find our process later
-    // in order to trigger the appropriate SIGHUP on config change.
-    let pid = std::process::id();
-    info!("process running in PID {pid}");
-    std::fs::write(args.pid_path, format!("{pid}\n")).context("writing PID to file")?;
-
```
Comment (conradludgate) [inline proxy/src/bin/local_proxy.rs:169]: leftover code I forgot to delete when switching over to the pid_file utility.

File: compute/vm-image-spec.yaml [hunk 7]
```diff
@@ -15,6 +15,10 @@ commands:
     user: postgres
     sysvInitAction: respawn
     shell: '/usr/local/bin/pgbouncer /etc/pgbouncer.ini'
+  - name: local-proxy
+    user: postgres
+    sysvInitAction: respawn
+    shell: '/usr/local/bin/local-proxy --config-path /etc/localproxy/config.json --pid-path /etc/localproxy/pid --http 0.0.0.0:10432'
```
Comment (tristan957) [inline compute/vm-image-spec.yaml:21]: Any chance we can use $XDG_RUNTIME_DIR for the PID path?

https://specifications.freedesktop.org/basedir-spec/latest/

File: compute_tools/src/config.rs [hunk 8]
```diff
@@ -139,3 +142,49 @@ where
 
     res
 }
+
+pub fn configure_local_proxy(local_proxy: &LocalProxySpec) -> Result<()> {
```
Comment (tristan957) [inline compute_tools/src/config.rs:146]: How do you feel about moving these functions into a local_proxy module?

File: compute_tools/src/compute.rs [hunk 9]
```diff
@@ -878,6 +879,11 @@ impl ComputeNode {
         // 'Close' connection
         drop(client);
 
+        if let Some(ref local_proxy) = spec.local_proxy_config {
+            info!("configuring local-proxy");
```
Comment (tristan957) [inline compute_tools/src/compute.rs:883]: Seems to be some discrepancies with `local-proxy` and `localproxy`

File: compute/vm-image-spec.yaml (hunk 7 above)
Comment (conradludgate) [reply compute/vm-image-spec.yaml:21]: I'm open to it, we don't seem to be using it anywhere else currently, so I'll have to check that the var is set for local-proxy and for compute-ctl.

File: compute_tools/src/compute.rs (hunk 9 above)
Comment (conradludgate) [reply compute_tools/src/compute.rs:883]: I'll switch everything over to snake case.

File: compute/vm-image-spec.yaml (hunk 7 above)
Comment (conradludgate) [reply compute/vm-image-spec.yaml]: Looks to be missing
```
root@compute-...:~# echo $XDG_RUNTIME_DIR

root@compute-...:~#
```

Comment (conradludgate) [reply compute/vm-image-spec.yaml]: (also not provided to compute_ctl env vars)

Comment (tristan957) [reply compute/vm-image-spec.yaml]: Hmm, that seems like a small issue, but not worth blocking on.

**Pull request merged**